libcaf_net/src/udp_datagram_socket.cpp

```diff
@@ -0,0 +1,137 @@
+#include "caf/detail/net_syscall.hpp"
+#include "caf/detail/socket_sys_aliases.hpp"
+#include "caf/detail/socket_sys_includes.hpp"
+#include "caf/expected.hpp"
+#include "caf/ip_endpoint.hpp"
+#include "caf/logger.hpp"
+#include "caf/net/socket_guard.hpp"
+#include "caf/span.hpp"
+
+namespace caf {
+namespace net {
+
+#ifdef CAF_WINDOWS
+
+error allow_connreset(udp_datagram_socket x, bool new_value) {
+  CAF_LOG_TRACE(CAF_ARG(x) << CAF_ARG(new_value));
+  DWORD bytes_returned = 0;
+  CAF_NET_SYSCALL("WSAIoctl", res, !=, 0,
+                  WSAIoctl(x.id, _WSAIOW(IOC_VENDOR, 12), &new_value,
+                           sizeof(new_value), NULL, 0, &bytes_returned, NULL,
+                           NULL));
+  return none;
+}
+
+#else // CAF_WINDOWS
+
+error allow_connreset(udp_datagram_socket x, bool) {
+  if (socket_cast<net::socket>(x) == invalid_socket)
+    return sec::socket_invalid;
+  // nop; SIO_UDP_CONNRESET only exists on Windows
+  return none;
+}
+
+#endif // CAF_WINDOWS
+
+expected<udp_datagram_socket> make_udp_datagram_socket(ip_endpoint& ep,
+                                                       bool reuse_addr) {
+  CAF_LOG_TRACE(CAF_ARG(ep));
+  auto addr = to_sockaddr(ep);
+  CAF_NET_SYSCALL("socket", fd, ==, invalid_socket_id,
+                  ::socket(addr.ss_family, SOCK_DGRAM, 0));
+  udp_datagram_socket sock{fd};
+  auto sguard = make_socket_guard(sock);
+  socklen_t len = (addr.ss_family == AF_INET) ? sizeof(sockaddr_in)
+                                              : sizeof(sockaddr_in6);
+  if (reuse_addr) {
+    int on = 1;
+    CAF_NET_SYSCALL("setsockopt", tmp1, !=, 0,
+                    setsockopt(fd, SOL_SOCKET, SO_REUSEADDR,
+                               reinterpret_cast<setsockopt_ptr>(&on),
+                               static_cast<socket_size_type>(sizeof(on))));
+  }
+  CAF_NET_SYSCALL("bind", err, !=, 0,
+                  ::bind(sock.id, reinterpret_cast<sockaddr*>(&addr), len));
+  CAF_NET_SYSCALL("getsockname", erro, !=, 0,
+                  getsockname(sock.id, reinterpret_cast<sockaddr*>(&addr),
+                              &len));
+  CAF_LOG_DEBUG(CAF_ARG(sock.id));
+  auto port = addr.ss_family == AF_INET
+                ? reinterpret_cast<sockaddr_in*>(&addr)->sin_port
+                : reinterpret_cast<sockaddr_in6*>(&addr)->sin6_port;
+  ep.port(ntohs(port));
+  return sguard.release();
+}
+
+variant<std::pair<size_t, ip_endpoint>, sec> read(udp_datagram_socket x,
+                                                  span<byte> buf) {
+  sockaddr_storage addr = {};
+  socklen_t len = sizeof(sockaddr_storage);
+  auto res = ::recvfrom(x.id, reinterpret_cast<socket_recv_ptr>(buf.data()),
+                        buf.size(), 0, reinterpret_cast<sockaddr*>(&addr),
+                        &len);
+  auto ret = check_udp_datagram_socket_io_res(res);
+  if (auto num_bytes = get_if<size_t>(&ret)) {
+    if (*num_bytes == 0)
+      CAF_LOG_INFO("Received empty datagram");
```

Comment:
The if-else tree only affects logging. I'd use the proper logging macros here instead:

```suggestion
      CAF_LOG_INFO_IF(*num_bytes == 0, "Received empty datagram");
```